homeassistant/components/google_calendar.py

```diff
@@ -0,0 +1,407 @@
+    # since we could just use the global and not assign it
+    import hashlib
+
+    calendar_hash = hashlib.sha224(
+        calendar['id'].encode('utf-8')).hexdigest()
+
+    if _CALENDARS.get(calendar_hash, None) is not None:
+        return
+
+    # seriously, FU pylint
+    _CALENDARS.update({calendar_hash: {
+        CONF_CAL_ID: calendar['id'],
+        CONF_NAME: calendar['summary'],
+        CONF_ENTITIES: [{
+            CONF_TRACK: track_new,
+            CONF_SEARCH: None,
+            CONF_NAME: calendar['summary']
+        }]
+    }})
+
+    update_config(
+        hass.config.path(YAML_DEVICES),
+        calendar_hash,
+        _CALENDARS[calendar_hash]
+    )
+
+    platform_discovered(hass, calendar_hash)
+
+
+def platform_discovered(hass, calendar_hash):
+    """Fire platform_discovered event."""
+    # Fire thermostat discovery event
+    hass.bus.fire(EVENT_PLATFORM_DISCOVERED, {
+        ATTR_SERVICE: DISCOVER_BINARY_SENSORS,
+        ATTR_DISCOVERED: _CALENDARS[calendar_hash]
+    })
+
+
+def get_next_event(hass, calendar_id, search=None):
+    """Get the latest data."""
+    from homeassistant.util import dt
+
+    service = get_calendar_service(hass)
+    params = dict(DEFAULT_GOOGLE_SEARCH_PARAMS)
+    params['timeMin'] = dt.utcnow().isoformat('T')
+    params['calendarId'] = calendar_id
+    if search:
+        params['q'] = search
+
+    result = service.events().list(**params).execute()
+
+    _LOGGER.error('Getting calendar data: [%s]', str(params))
+
+    items = result.get('items', [])
+    return items[0] if len(items) == 1 else None
+
+
+def get_calendar_service(hass):
+    """Get the calendar service from the storage file token."""
+    # pylint disable=no-name-in-module,import-error
+    import httplib2
+    from oauth2client.file import Storage
+    from googleapiclient import discovery
+    token_file = hass.config.path(TOKEN_FILE)
+    credentials = Storage(token_file).get()
+    http = credentials.authorize(httplib2.Http())
+    service = discovery.build('calendar', 'v3', http=http)
+    return service
+
+
+def setup_scanner_platform(hass, config):
+    """Setup the scanner to look for new calendars."""
+    from homeassistant.util import convert
+    from homeassistant.helpers.event import track_utc_time_change
+    interval = convert(config.get(CONF_SCAN_INTERVAL), int,
+                       DEFAULT_CONF_SCAN_INTERVAL)
+
+    def calendar_tracker_scan(now):
+        """Call up service and get list of calendars.
+
+        Then call see_calendar() on each.
+
+        """
+        # pylint disable=unused-argument
+        hass.services.call(DOMAIN, SERVICE_SCAN_CALENDARS, None)
+
+    track_utc_time_change(hass,
+                          calendar_tracker_scan,
+                          minute=range(0, 60, interval),
+                          second=0)
+
+
+def load_config(path):
+    """Load the google_calendar_devices.yaml."""
+    import os
```

Comment:
Imports for Python built-in modules or home assistant modules can be at the top of the document.